File: src/main/java/tk/hiddenname/smarthome/service/hardware/manager/SensorManager.java
package tk.hiddenname.smarthome.service.hardware.manager;

import lombok.AllArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import tk.hiddenname.smarthome.entity.hardware.GPIO;
import tk.hiddenname.smarthome.entity.hardware.Sensor;
import tk.hiddenname.smarthome.entity.signal.SignalType;
import tk.hiddenname.smarthome.exception.GPIOBusyException;
import tk.hiddenname.smarthome.exception.PinSignalSupportException;
import tk.hiddenname.smarthome.exception.SignalTypeNotFoundException;
import tk.hiddenname.smarthome.service.database.SensorDatabaseService;
import tk.hiddenname.smarthome.service.hardware.impl.GPIOService;
import tk.hiddenname.smarthome.service.hardware.impl.digital.input.DigitalSensorServiceImpl;

@Component
@AllArgsConstructor
public class SensorManager {

    private static final Logger log = LoggerFactory.getLogger(SensorManager.class);

    private final DigitalSensorServiceImpl digitalService;
    private final SensorDatabaseService service;

    public void create(Sensor sensor) throws PinSignalSupportException, GPIOBusyException {
        log.debug("Creating device " + sensor.toString());
        GPIO gpio = sensor.getGpio();
        getService(gpio.getType()).save(
                sensor.getId(),
                gpio.getGpio(),
                sensor.isReverse()
        );
    }

    public void delete(Sensor sensor) {
        log.debug("Deleting sensor " + sensor.toString());
        GPIO gpio = sensor.getGpio();
        getService(gpio.getType()).delete(
                sensor.getId()
        );
    }

    public void loadSensors() {
        for (Sensor sensor : service.getAll()) {
            try {
                create(sensor);
            } catch (PinSignalSupportException | GPIOBusyException e) {
                log.warn(e.getMessage());
            }
        }
    }

    private GPIOService getService(SignalType type) {
        if (type == SignalType.DIGITAL) {
            return digitalService;
        }
        SignalTypeNotFoundException e = new SignalTypeNotFoundException(type.toString());
        log.warn(e.getMessage());
        throw e;
    }
}

File: src/main/kotlin/tk/hiddenname/smarthome/service/hardware/manager/SensorManager.kt
package tk.hiddenname.smarthome.service.hardware.manager

import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import tk.hiddenname.smarthome.exception.not_specified.GpioNotSpecifiedException
import tk.hiddenname.smarthome.exception.exist.GpioPinBusyException
import tk.hiddenname.smarthome.exception.not_specified.SignalTypeNotSpecifiedException
import tk.hiddenname.smarthome.exception.support.PinSignalSupportException
import tk.hiddenname.smarthome.exception.support.UnsupportedSignalTypeException
import tk.hiddenname.smarthome.model.hardware.Sensor
import tk.hiddenname.smarthome.model.signal.SignalType
import tk.hiddenname.smarthome.service.database.SensorDatabaseService
import tk.hiddenname.smarthome.service.hardware.impl.GPIOService
import tk.hiddenname.smarthome.service.hardware.impl.digital.input.DigitalSensorServiceImpl
import tk.hiddenname.smarthome.utils.gpio.GpioManager

@Component
class SensorManager(private val service: SensorDatabaseService,
                    private val digitalService: DigitalSensorServiceImpl,
                    private val gpioManager: GpioManager) {

    private val log = LoggerFactory.getLogger(SensorManager::class.java)

    @Throws(PinSignalSupportException::class, GpioPinBusyException::class)
    fun register(sensor: Sensor) {
        log.debug("Creating sensor $sensor")
        sensor.gpio ?: throw GpioNotSpecifiedException()
        getService(sensor.gpio.signalType).save(
                sensor.id,
                sensor.gpio,
                sensor.signalInversion
        )
    }

    fun unregister(sensor: Sensor) {
        log.debug("Deleting sensor $sensor")
        sensor.gpio ?: throw GpioNotSpecifiedException()
        getService(sensor.gpio.signalType).delete(
                sensor.id
        )
    }

    fun loadSensors() {
        service.getAll().forEach {
            try {
                register(it)
            } catch (e: PinSignalSupportException) {
                log.warn(e.message)
            } catch (e: GpioPinBusyException) {
                log.warn(e.message)
            }
        }
    }

    @Suppress("DuplicatedCode")
    fun changeSignalType(sensor: Sensor, newSignalType: SignalType?) {
        newSignalType ?: throw SignalTypeNotSpecifiedException()
        sensor.gpio ?: throw GpioNotSpecifiedException()
        if (gpioManager.validate(sensor.gpio.gpioPin, newSignalType, sensor.gpio.pinMode)) {
            unregister(sensor)
            sensor.gpio.signalType = newSignalType
            register(sensor)
        }
    }

    private fun getService(type: SignalType?): GPIOService {
        return when (type) {
            SignalType.DIGITAL -> digitalService
            else -> throw UnsupportedSignalTypeException(type!!)
        }
    }
}